src/binder_v2/mod.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::vec::Vec;
+
+use egg::Id;
+use itertools::Itertools;
+
+use crate::catalog::{RootCatalog, TableRefId, DEFAULT_DATABASE_NAME, DEFAULT_SCHEMA_NAME};
+use crate::parser::*;
+use crate::planner::{Expr as Node, ExprAnalysis, RecExpr};
+use crate::types::{DataTypeKind, DataValue};
+
+mod expr;
+mod select;
+mod table;
+
+pub use self::expr::*;
+pub use self::select::*;
+pub use self::table::*;
+
+pub type Result<T = Id> = std::result::Result<T, BindError>;
+
+/// The error type of bind operations.
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum BindError {
+    #[error("invalid database {0}")]
+    InvalidDatabase(String),
+    #[error("invalid schema {0}")]
+    InvalidSchema(String),
+    #[error("invalid table {0}")]
+    InvalidTable(String),
+    #[error("invalid column {0}")]
+    InvalidColumn(String),
+    #[error("duplicated table {0}")]
+    DuplicatedTable(String),
+    #[error("duplicated column {0}")]
+    DuplicatedColumn(String),
+    #[error("duplicated alias {0}")]
+    DuplicatedAlias(String),
+    #[error("invalid expression: {0}")]
+    InvalidExpression(String),
+    #[error("not nullable column: {0}")]
+    NotNullableColumn(String),
+    #[error("binary operator types mismatch: {0} != {1}")]
+    BinaryOpTypeMismatch(String, String),
+    #[error("ambiguous column: {0}")]
+    AmbiguousColumn(String),
+    #[error("invalid table name: {0:?}")]
+    InvalidTableName(Vec<Ident>),
+    #[error("SQL not supported")]
+    NotSupportedTSQL,
+    #[error("invalid SQL")]
+    InvalidSQL,
+    #[error("cannot cast {0:?} to {1:?}")]
+    CastError(DataValue, DataTypeKind),
+    #[error("{0}")]
+    BindFunctionError(String),
+}
+
+/// The binder resolves all expressions referring to schema objects such as
+/// tables or views with their column names and types.
+pub struct Binder {
+    egraph: egg::EGraph<Node, ExprAnalysis>,
+    catalog: Arc<RootCatalog>,
+    contexts: Vec<Context>,
+}
+
+/// The context of binder execution.
+#[derive(Debug, Default)]
+struct Context {
+    /// Mapping table name to its ID.
+    tables: HashMap<String, TableRefId>,
```

Comment:
I'll prefer make Binder stateless in the future. Will do the refactor later.

Reply:
Do you mean to make the context an analysis data associated with each node?
There would be more overhead to maintain the context for each node, but significantly reduce the possibility to get things wrong. So I would +1 for this proposal.

Reply:
I think we are on the same page. e.g. if we want to resolve a column name, we look into the BoundTableRef (or the context bound to the table ref) instead of looking at the global context.